File: problem560.py
"""

nums = [1,1,1], k=2

nums = [1,2,3], k=3

complement = [-2], currsum=1, res = 0
complement = [-2, 0, ], currsum=3, res = 1
complement = [0, -2], currsum=3, res = 2


nums = [1,1,2,-1,-5,2,3], target=3
i=0, comp=[2,], sum=1, res=0
i=1, comp=[2,1],sum=2, res=0
i=2, comp=[2,1,-1], sum=4, res=0
i=3, comp=[2,1,-1,0], sum=3, res=1,
i=4, comp=[2,1,-1,5], sum=-2, res=1
i=5, comp=[2,1,-1,5], sum=0, res=2,
i=6, comp=[0,2,1,-1,5], sum=3, res=2,


"""

class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        complement = defaultdict(int)
        complement[0]=1
        result=0
        currsum=0
        for n in nums:
            currsum+=n
            result+=complement[currsum-k]
            complement[currsum]+=1
            
        return result
        
#Time: O(N)
#Space: O(N)
# https://replit.com/@vokoshyv/DearExpensiveLevel#index.js

File: problem560.cs
/*

[1 2 3 4 -7 3 3], k=3, complement={0:1}
i=0, currsum=1, complement={0:1, 1:0}, result=0
i=1, currsum=3, complement={0:1, 1:0, 3:1}, result=1
i=2, currsum=6, complement={0:1, 1:0, 3:1, 6:1}, result=2
i=3, currsum=10, complement={0:1, 1:0, 3:1, 6:1, 10:1}, result=2
i=4, currsum=3, complement={0:1, 1:0, 3:2, 6:1, 10:1}, result=3
i=5, currsum=6, complement={0:1, 1:0, 3:2, 6:2, 10:1}, result=5
i=6, currsum=9, complement={0:1, 1:0, 3:2, 6:2, 10:1, 9:1}, result=7

*/

public class Solution {
    public int SubarraySum(int[] nums, int k) {
        int currsum=0;
        int result=0;
        
        Dictionary<int, int> complement=new Dictionary<int, int>(20000);
        complement[0]=1;
        for(int i=0; i<nums.Length; i++){
            currsum+=nums[i];
            if(!complement.ContainsKey(currsum-k)){
                complement[currsum-k]=0;
            }
            result+=complement[currsum-k];
            if(!complement.ContainsKey(currsum)){
                complement[currsum]=0;
            }
            complement[currsum]++;
        }
        
        return result;
    }
}

//time: O(N)
//space: O(N)